hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngrade.java

```diff
@@ -962,6 +964,17 @@ public void testPayloadUpgradeDowngrade(String tableType, RecordMergeMode record
         "Table should be downgraded to version 6 for payload: " + payloadType);
     validateDataConsistency(expectedData, metaClientV6, "after v9->v6 downgrade for " + payloadType);
 
+    // Perform read-optimized query and validate data
+    // (Note compaction seems to have already occured at this point)
+    // meaning ro_query should have same results as snapshot query
+    LOG.info("Performing read-optimized query for payload: {}", payloadType);
+    Dataset<Row> readOptimizedData = sqlContext().read()
+        .format("hudi")
+        .option("hoodie.datasource.query.type", "read_optimized")
+        .load(metaClientV6.getBasePath().toString());
+
+    assertEquals(6, readOptimizedData.count(),
```

Comment:
Do ro query after upgrade and then do same after downgrade. Ensure data frame validation happens after both.